Implement fog

JayFoxRox opened the pull request:
This implements fog mode 5.
(Known fog modes are None = 0; Fog = 5; Gas = 7. `master` treats everything like mode 0)

---

The fog mode is fully implemented in SW rasterizer and HW renderer.
The fog was hardware tested (against all logic op modes, blend modes, RGBA write masks, Z-Buffer / W-Buffer, LUT overflows / underflows, LUT index wrapping) and it's been found that this fog mode is like a final-combiner for the R, G, B channels. There is no alpha fog or dual source blending.

The overflow of the fog_lut_offset was confirmed on hardware. I didn't want to change the size of the BitField though because there might be reasons why 3dbrew lists it as 16 bit and we can't read back the actual size of the hardware register.

There are a handful of tiny accuracy problems remaining but we can't solve them until we figure out how exactly the 3DS interpolates the depth etc. I've mentioned this in the SW rasterizer code (HW is inaccurate anyway).

There is only one difference I could find between 3DS fog and this implementation: the fog LUT is cleared to zero in Citra while the 3DS seems to keep random values in it = it's not initialized for the process.
I believe the fog index is still initialized to 0 (somewhere unknown - `FINALIZE` maybe?) because Luigi's Mansion Dark Moon does fog LUT writes (in batches of 128 = fog LUT size) without ever explicitly setting the index (Note that it doesn't ever seem to use fog mode 5 either).

There are no known problems or missing functionality for this fog mode.

---

During development, I was also asked for a "Disable fog" enhancement option in the configuration. 
- I'll leave that to someone else.
(Fog is often used for graphics effects and gives a sense of scale so I think disabling it would be pretty stupid and not an enhancement. If anybody decides to write such an enhancement they should probably just set a falloff curve for the fog instead so that it can be shifted back slightly)

---

This fixes the OoT fog and more..

Comparison screenshots:

[Citra `master`](http://imgur.com/a/AO6t9):
![Zelda (Citra master)](http://i.imgur.com/wBifF9w.png)

Citra `fog`:
![Zelda (Citra PR)](http://i.imgur.com/ho9An1Z.png)

[Real console](http://imgur.com/a/jHhA5) (Thanks to @FenrisulfrX):
![Zelda (Console)](http://i.imgur.com/nawyMqT.png)

More Screenshots by @Leo626: [Majora's Mask - Mountain Village](http://imgur.com/a/Mtw3w), [Mario Kart 7 - DK Pass](http://imgur.com/a/TrXaY), [Super Mario 3D Land - S6-5](http://imgur.com/a/ZjF1a). Also regression tested all stages of Kingdom Hearts 3D (with SW / HW renderer) and everything works.

---

Massive thanks to @Leo626 for reporting various major issues during review on my repo.
Also thanks to @wwylele, **Tohka**, @einstein95 and @FenrisulfrX (+ everybody I forgot) who ran the hardware tests for me.

---

TODO:
- [x] ~~hw-test zflip~~
- [ ] hw-test wbuffer depth scale
- [x] ~~Create a hw test~~
- [x] ~~Fix regression in Kingdom Hearts 3D (reported by Leo626)~~
- [x] ~~Wait for w-buffer PR to go through~~
- [x] ~~Rebase onto master~~
- [x] ~~Split into sw / hw~~
- [x] ~~Get this running~~
- [x] ~~Implement z-flip~~
- [x] ~~Fix PR description~~
- [x] ~~Fix weird bug about fog breaking on SW->HW~~
- [x] ~~Comparison to  (Thanks to  for these screenshots)~~

JayFoxRox commented:
Totally bugs out (very dense fog) in OoT when switching HW/SW often. Problem happens only in HW renderer and can be fixed by talking to navi. Will happen again.

JayFoxRox committed:
Pica: Add fog state

JayFoxRox committed:
Rasterizer: Implement fog

JayFoxRox commented on src/video_core/renderer_opengl/gl_shader_gen.cpp:
```diff
@@ -598,10 +602,12 @@ vec4 secondary_fragment_color = vec4(0.0);
-    // Append fog blending
-    if (state.fog_mode == Regs::FogMode::Fog) {
+    if (VideoCore::g_fog_enabled && state.fog_mode == Regs::FogMode::Fog) {
+#if 1
```
NOPE..

JayFoxRox commented:
This needs a review but is otherwise good to go.

JayFoxRox commented:
Looks like I accidently added includes again which were removed in some new PR

JayFoxRox commented on src/video_core/renderer_opengl/gl_rasterizer.h:
```diff
@@ -423,4 +433,7 @@ class RasterizerOpenGL : public VideoCore::RasterizerInterface {
     std::array<std::array<GLvec4, 256>, 6> lighting_lut_data;
+
+    OGLTexture fog_lut;
+    std::array<GLuint, 128> fog_lut_data;
```
`std::array<GLuint, 128> fog_lut_data{};` once my other PR has been merged

Leo626 commented:
Hi @JayFoxRox I get some errors while compiling with MSVC. [First set](http://i.imgur.com/evyhhau.png) pertain to 'g_fog_enabled': is not a member of 'VideoCore' and 'g_fog_enabled': undeclared identifier while the [second set](http://i.imgur.com/VRWHhp4.png) is about Element 1, 2 and 3: conversion from 'u32' to 'const u8 &' requires a narrowing conversion. For the narrowing conversion error for example, I see it points to [L1188](https://github.com/JayFoxRox/citra/blob/fog/src/video_core/rasterizer.cpp#L1188) of rasterizer,cpp. Full log from MSVC provided [here](http://clipboard.space/clip/xkMrERk6WLcoyY6pqLKZ).

Let me know if I'm not providing enough info. If you think this is not worth your time, then feel free to ignore and delete it.

JayFoxRox commented:
Should be fixed

Leo626 commented:
Compiles without any errors and works. Feel free to delete my comments.

JayFoxRox commented:
Rebased

Leo626 commented:
There appears to be a regression with Kingdom Hearts 3D. [Models lack textures](http://imgur.com/a/mePjq) while some areas are covered in [thick fog](http://imgur.com/a/XFPya). Occurs in cutscenes and gameplay. World Map and Tron level appear unaffected but slightly darker.

JayFoxRox commented:
Interesting. I'll try to get my hands on that game or another game with those kind of issues.
I also plan to write hardware tests soon(tm), so I might be able to fix it then too.

JayFoxRox commented:
There also seems to be a bug in how the LUT is used:

```
fog 1: 2047 0
fog 2: 2047 0
fog 4: 2047 0
fog 5: 2047 0
fog 7: 2047 0
fog 8: 2047 0
fog 10: 2047 0
fog 11: 2047 0
fog 13: 2047 0
fog 14: 2047 0
fog 16: 2047 0
fog 17: 2047 0
fog 19: 2047 0
fog 20: 2047 0
fog 22: 2047 0
fog 23: 2047 0
fog 25: 2047 0
fog 26: 2047 0
fog 28: 2047 0
fog 29: 2047 0
fog 31: 2047 0
fog 32: 2047 0
fog 34: 2047 0
fog 35: 2047 0
fog 37: 2047 0
fog 38: 2047 0
fog 40: 2047 0
fog 41: 2047 0
fog 43: 2047 0
fog 44: 2047 0
fog 46: 2047 0
fog 47: 2047 0
fog 49: 2047 0
fog 50: 2047 0
fog 52: 2047 0
fog 53: 2047 0
fog 55: 2047 0
fog 56: 2047 0
fog 58: 2047 0
fog 59: 2047 0
fog 61: 2047 0
fog 62: 2047 0
fog 64: 2047 0
fog 65: 2047 0
fog 67: 2047 0
fog 68: 2047 0
fog 70: 2047 0
fog 71: 2047 0
fog 73: 2047 0
fog 74: 2047 0
fog 76: 2047 0
fog 77: 2047 0
fog 79: 2047 0
fog 80: 2047 0
fog 82: 2047 0
fog 83: 2047 0
fog 85: 2047 0
fog 86: 2047 0
fog 88: 2047 0
fog 89: 2047 0
fog 91: 2047 0
fog 92: 2047 0
fog 94: 2047 0
fog 95: 2047 0
fog 97: 2047 0
fog 98: 2047 0
fog 100: 2047 0
fog 101: 2047 0
fog 103: 2047 0
fog 104: 2047 0
fog 106: 2047 0
fog 107: 2047 0
fog 109: 2047 0
fog 110: 2047 0
fog 112: 2047 0
fog 113: 2047 0
fog 115: 2047 0
fog 116: 2047 0
fog 118: 2047 0
fog 119: 2047 0
fog 121: 2047 0
fog 122: 2047 0
fog 124: 2047 0
fog 125: 2047 0
fog 127: 2047 0
fog 128: 2047 0
fog 130: 2047 0
fog 131: 2047 0
fog 133: 2047 0
fog 134: 2047 0
fog 136: 2047 0
fog 137: 2047 0
fog 139: 2047 0
fog 140: 2047 0
fog 142: 2047 0
fog 143: 2047 0
fog 145: 2047 0
fog 146: 2047 0
fog 148: 2047 0
fog 149: 2047 0
fog 151: 2047 0
fog 152: 2047 0
fog 154: 2047 0
fog 155: 2047 0
fog 157: 2047 0
fog 158: 2047 0
fog 160: 2047 0
fog 161: 2047 0
fog 163: 2047 0
fog 164: 2047 0
fog 166: 2047 0
fog 167: 2047 0
fog 169: 2047 0
fog 170: 2047 0
fog 172: 2047 0
fog 173: 2047 0
fog 175: 2047 0
fog 176: 2047 0
fog 178: 2047 0
fog 179: 2047 0
fog 181: 2047 0
fog 182: 2047 0
fog 184: 2047 0
fog 185: 2047 0
fog 187: 2047 0
fog 188: 2047 0
fog 190: 2047 0
fog 191: 2047 0
fog 193: 2047 0
fog 194: 2047 0
fog 196: 2047 0
fog 197: 2047 0
fog 199: 2047 0
fog 200: 2047 0
fog 202: 2047 0
fog 203: 2047 0
fog 205: 2047 0
fog 206: 2047 0
fog 208: 2047 0
fog 209: 2047 0
fog 211: 2047 0
fog 212: 2047 0
fog 214: 2047 0
fog 215: 2047 0
fog 217: 2047 0
fog 218: 2047 0
fog 220: 2047 0
fog 221: 2047 0
fog 223: 2047 0
fog 224: 2047 0
fog 226: 2047 0
fog 227: 2047 0
fog 229: 2047 0
fog 230: 2047 0
fog 232: 2047 0
fog 233: 2047 0
fog 235: 2047 0
fog 236: 2047 0
fog 238: 2047 0
fog 239: 2047 0
fog 241: 2047 0
fog 242: 2047 0
fog 244: 2047 0
fog 245: 2047 0
fog 247: 2047 0
fog 248: 2047 0
fog 250: 2047 0
fog 251: 2047 0
fog 253: 2047 0
fog 254: 2047 0
fog 256: 2047 0
fog 257: 2047 0
fog 259: 2047 0
fog 260: 2047 0
fog 262: 2047 0
fog 263: 2047 0
fog 265: 2047 0
fog 266: 2047 0
fog 268: 2047 0
fog 269: 2047 0
fog 271: 2047 0
fog 272: 2047 0
fog 274: 2047 0
fog 275: 2047 0
fog 277: 2047 0
fog 278: 2047 0
fog 280: 2047 0
fog 281: 2047 0
fog 283: 2047 0
fog 284: 2047 0
fog 286: 2047 0
fog 287: 2047 0
fog 289: 2047 0
fog 290: 2047 0
fog 292: 2047 0
fog 293: 2047 0
fog 295: 2047 0
fog 296: 2047 0
fog 298: 2047 0
fog 299: 2047 0
fog 301: 2047 0
fog 302: 2047 0
fog 304: 2047 0
fog 305: 2047 0
fog 307: 2047 0
fog 308: 2047 0
fog 310: 2047 0
fog 311: 2047 0
fog 313: 2047 0
fog 314: 2047 0
fog 316: 2047 0
fog 317: 2047 0
fog 319: 2047 0
fog 320: 2047 0
fog 322: 2047 0
fog 323: 2047 0
fog 325: 2047 0
fog 326: 2047 0
fog 328: 2047 0
fog 329: 2047 0
fog 331: 2047 0
fog 332: 2047 0
fog 334: 2047 0
fog 335: 2047 0
fog 337: 2047 -1
fog 338: 2047 -1
fog 340: 2047 -1
fog 341: 2047 -1
fog 343: 2047 -1
fog 344: 2047 -1
fog 346: 2047 -1
fog 347: 2047 -1
fog 349: 2047 -1
fog 350: 2047 -1
fog 352: 2046 -1
fog 353: 2046 -1
fog 355: 2046 -1
fog 356: 2046 -1
fog 358: 2045 -1
fog 359: 2045 -1
fog 361: 2045 -1
fog 362: 2045 -1
fog 364: 2044 -2
fog 365: 2044 -2
fog 367: 2043 -2
fog 368: 2043 -2
fog 370: 2041 -3
fog 371: 2041 -3
fog 373: 2039 -4
fog 374: 2039 -4
fog 376: 2035 -8
fog 377: 2035 -8
fog 379: 2028 -22
fog 380: 2028 -22
fog 382: 2007 -1565
fog 383: 2007 -1565
```
(3 lines per entry, 1 hidden. OoT splash scene)

Looks like it should be `value - difference * fog_f` or even `next_value - value * difference`

JayFoxRox commented:
@Leo626: Should be fixed, can you regression test this on like.. as many games as possible?
Please test both, sw rasterizer and hw.
Ideally every game should have working fog (just like a real 3DS, no differences hopefully) and no regressions.

(I also rebased #12 which is based on this branch and supports fog debugging (must open the pica combiner debugger widget). So in case of errors I might ask you for data provided by that)

Leo626 commented:
The issue with KH3D is fixed. Visited each world and played the tutorial on both hw and sw-rasterizer. From what I've tested so far, fog is very noticeable in these levels/stages:

- Majora's Mask - Termina Field and Snowhead
- Mario Kart 7 - DK Pass
- Super Mario 3D Land - W1-3, W3-4 and S6-5

Super Mario 3D Land looks glitched (aside from the broken effects from imm-mode2) in S6-5 and probably the other levels in sw-rasterizer. Pillars are visible through platform.

[Without fix-batching (hw on top and sw below)](http://imgur.com/a/rNDHH)
[With fix-batching (hw on top and sw below)](http://imgur.com/a/yK0kG)

JayFoxRox commented:
That's probably another issue, unrelated to fog (or maybe fog on alpha?!)
Also how does that world look like without fog in SW-rasterizer? Are the columns also in front of the other stuff?

Leo626 commented:
[Without fog in sw-rasterizer.](http://i.imgur.com/sgNFhTu.png)

JayFoxRox commented:
There is a bug with fog on logic-op in SM3DL intro. I'll probably have to hw-test that but I'll also add it to the combiner debugger

JayFoxRox commented:
@Leo626 which other fixes do you use? SM3DL crashes with unsupported lighting mode for me

```
[ 238.317429] HW.GPU <Critical> video_core/renderer_opengl/gl_shader_gen.cpp:operator():402: Unknown lighting LUT input 5
```

causes

```
[ 238.319899] Debug <Critical> video_core/renderer_opengl/gl_rasterizer.cpp:operator():930: Assertion Failed!
Uniform block size did not match!
```

=> assert failed => exit

Leo626 commented:
I didn't use any other other fixes. I just used fog 247b0cf for it. For me, I get this during the intro:

`[ 216.794043] Render.OpenGL <Error> video_core\renderer_opengl\gl_rasterizer.cpp:RasterizerOpenGL::DrawTriangles:223: Fog on LogicOp not implemented in hardware renderer
`

and it crashes when I start the first level on a new game. It crashes with this in the log:

`[ 216.946170] Debug <Critical> video_core\renderer_opengl\gl_rasterizer.cpp:RasterizerOpenGL::SetShader::<lambda_aff6bd6d03ba202a2f14100d22fddd6e>::operator ():930: Assertion Failed!
Uniform block size did not match!`

Latest nightly doesn't crash. Sorry if I'm not being helpful.

JayFoxRox commented:
@Leo626 can you share your savegame? i don't feel like playing through all other worlds to reach the bug at the moment :P

Leo626 commented:
The first save file is right on S6-5. Rename to 00053F00 if using European version.
[00054000.zip](https://github.com/JayFoxRox/citra/files/273433/00054000.zip)

JayFoxRox commented:
Note to self: Can be fixed by disabling fog on logic ops.. probably isn't done on hw either! needs hw-test

Leo626 commented:
Luigi's Mansion Dark Moon crashes when trying to boot.

```
[ 280.076043] Debug <Critical> video_core\command_processor.cpp:Pica::CommandProcessor::WritePicaReg::<lambda_c78dfa2ac9fb41a4dd6e808628d4f507>::operator ():431: Assertion Failed!
fog_lut_offset exceeded maximum value of 127!
```

JayFoxRox commented:
FIXME: Do some alpha test crap. Fog seems to be part of the TEV stages as a final combiner. That's pretty unfancy for such a modern GPU. I'll take it though

JayFoxRox committed:
OpenGL: Implement fog

JayFoxRox commented on src/video_core/rasterizer.cpp:
```diff
@@ -833,6 +854,28 @@ static void ProcessTriangleInternal(const Shader::OutputVertex& v0,
                     continue;
             }
 
+            // Apply fog combiner
```
Add a note that this might be inaccurate!

JayFoxRox commented:
@Leo626 can you create some comparison screenshots from real 3DS compared to citra master compared to Citra HW / SW. I'll put the PR live later and I guess it would help the review process to provide eyecandy :)
Also we'll be able to catch the last regressions

Leo626 commented:
@JayFoxRox will do.

Leo626 commented:
- [Majora's Mask - Mountain Village](http://imgur.com/a/Mtw3w)
- [Mario Kart 7 - DK Pass](http://imgur.com/a/TrXaY)
- [Super Mario 3D Land - S6-5](http://imgur.com/a/ZjF1a)

JayFoxRox commented:
This is ready to go as far as I'm concerned. However, I have too many open PRs and will be pretty busy the next weeks.
I'll just hold this back until more of my other stuff is merged.

Also I still have to edit in the OoT eyecandy into the PR description + review myself (description + code quality).

uberhalit commented:
According to some users from [gbatemp](https://gbatemp.net/threads/citra-new-3ds-emulator.365154/page-214#post-6375354) this PR breaks hardware rendering in AMD graphic cards.

JayFoxRox commented:
Yes, I'm aware of said post. However, I'm not convinced that it's a problem with my code.
\- I suspect a driver bug or merge error.

Unless I get proper bug reports (system / driver information, used master (best to just use this entire branch), full log, maybe even apitrace) I have no issues merging this. I've confirmed functionality on a handful of GL implementations without any warnings.

uberhalit commented:
Fair enought.
I tested it on 2 mobile Nvidia devices and so far I encountered no problems.

JayFoxRox committed:
OpenGL: Avoid undefined behaviour for UNIFORM_BLOCK_DATA_SIZE

JayFoxRox commented:
GitHub seems to be really buggy at the moment, but theoretically the AMD issue should be fixed now.

mailwl committed:
gsp::gpu: Reset g_thread_id in UnregisterInterruptRelayQueue

archshift committed:
Remove unused and bitrotted "controller config" files

bunnei committed:
Merge pull request #1863 from mailwl/gpu-threadid-reset

gsp::gpu: Reset g_thread_id in UnregisterInterruptRelayQueue

lioncash committed:
service: Add other DLP services

Specifically, dlp::CLNT and dlp::FKCL

Moves them to their own folder like with other services.

bunnei committed:
Merge pull request #1884 from lioncash/dlp

service: Add other DLP services

bunnei committed:
Merge pull request #1873 from archshift/remove-config

Remove unused and bitrotted "controller config" files

JayFoxRox commented:
Moved to main project already.
